internal/controller/nnfsystemstorage_controller.go

```diff
@@ -0,0 +1,632 @@
+		if nnfSystemStorage.Spec.SystemConfiguration.Kind != reflect.TypeOf(dwsv1alpha2.SystemConfiguration{}).Name() {
+			return nil, dwsv1alpha2.NewResourceError("system configuration is not of kind '%s'", reflect.TypeOf(dwsv1alpha2.SystemConfiguration{}).Name()).WithFatal()
+		}
+
+		systemConfiguration.ObjectMeta = metav1.ObjectMeta{
+			Name:      nnfSystemStorage.Spec.SystemConfiguration.Name,
+			Namespace: nnfSystemStorage.Spec.SystemConfiguration.Namespace,
+		}
+	} else {
+		systemConfiguration.ObjectMeta = metav1.ObjectMeta{
+			Name:      "default",
+			Namespace: corev1.NamespaceDefault,
+		}
+	}
+
+	if err := r.Get(ctx, client.ObjectKeyFromObject(systemConfiguration), systemConfiguration); err != nil {
+		return nil, dwsv1alpha2.NewResourceError("could not get systemconfiguration '%v'", client.ObjectKeyFromObject(systemConfiguration)).WithError(err)
+	}
+
+	return systemConfiguration, nil
+}
+
+// Get the StorageProfile specified in the spec. We don't look for the default profile, a profile must be
+// specified in the NnfSystemStorage spec, and it must be marked as pinned.
+func (r *NnfSystemStorageReconciler) getStorageProfile(ctx context.Context, nnfSystemStorage *nnfv1alpha1.NnfSystemStorage) (*nnfv1alpha1.NnfStorageProfile, error) {
+	if nnfSystemStorage.Spec.StorageProfile == (corev1.ObjectReference{}) {
+		return nil, dwsv1alpha2.NewResourceError("StorageProfile must be specified").WithFatal()
+	}
+
+	if nnfSystemStorage.Spec.StorageProfile.Kind != reflect.TypeOf(nnfv1alpha1.NnfStorageProfile{}).Name() {
+		return nil, dwsv1alpha2.NewResourceError("StorageProfile is not of kind '%s'", reflect.TypeOf(nnfv1alpha1.NnfStorageProfile{}).Name()).WithFatal()
+	}
+
+	storageProfile := &nnfv1alpha1.NnfStorageProfile{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nnfSystemStorage.Spec.StorageProfile.Name,
+			Namespace: nnfSystemStorage.Spec.StorageProfile.Namespace,
+		},
+	}
+
+	if err := r.Get(ctx, client.ObjectKeyFromObject(storageProfile), storageProfile); err != nil {
+		return nil, dwsv1alpha2.NewResourceError("could not get StorageProfile '%v'", client.ObjectKeyFromObject(storageProfile)).WithError(err)
+	}
+
+	return storageProfile, nil
+}
+
+// Create a Servers resource with one allocation on each Rabbit. If the IncludeRabbits array is not
+// empty, only use those Rabbits. Otherwise, use all the Rabbits in the SystemConfiguration resource except
+// those specified in the ExcludeRabbits array.
+func (r *NnfSystemStorageReconciler) createServers(ctx context.Context, nnfSystemStorage *nnfv1alpha1.NnfSystemStorage) error {
+	log := r.Log.WithValues("NnfSystemStorage", client.ObjectKeyFromObject(nnfSystemStorage))
+
+	// Create a list of Rabbits to use
+	rabbitList := []string{}
+
+	if len(nnfSystemStorage.Spec.IncludeRabbits) != 0 {
+		if len(nnfSystemStorage.Spec.ExcludeRabbits) != 0 {
+			return dwsv1alpha2.NewResourceError("IncludeRabbits and ExcludeRabbits can not be used together").WithFatal()
+		}
```

Comment:
Should this be in a webhook?

Reply:
I like the idea, but it's probably not that useful for a resource that is going to see such little use. Maybe I can add it as a follow on so we can get the base functionality out in a release sooner?